src/storage-operators/src/s3_oneshot_sink/parquet.rs

```diff
@@ -7,28 +7,367 @@
+    /// and to make it easier to take ownership when calling in spawned
+    /// tokio tasks (to avoid doing I/O in the surrounding timely context).
+    active_file: Option<ParquetFile>,
+    /// Upload and buffer params
+    params: CopyToParameters,
+}
 
 impl CopyToS3Uploader for ParquetUploader {
     fn new(
-        _sdk_config: SdkConfig,
-        _connection_details: S3UploadInfo,
-        _sink_id: &GlobalId,
-        _batch: u64,
+        sdk_config: SdkConfig,
+        connection_details: S3UploadInfo,
+        sink_id: &GlobalId,
+        batch: u64,
+        params: CopyToParameters,
     ) -> Result<ParquetUploader, anyhow::Error> {
-        anyhow::bail!("Parquet uploading is not yet implemented")
+        if params.parquet_row_group_ratio > 100 {
+            anyhow::bail!("parquet_row_group_ratio must be <= 100");
+        }
+        if params.arrow_builder_buffer_ratio < 100 {
+            anyhow::bail!("arrow_builder_buffer_ratio must be >= 100");
+        }
+        let row_group_size_bytes =
+            connection_details.max_file_size * u64::cast_from(params.parquet_row_group_ratio) / 100;
+        let arrow_builder_buffer_bytes =
+            row_group_size_bytes * u64::cast_from(params.arrow_builder_buffer_ratio) / 100;
+
+        match connection_details.format {
+            S3SinkFormat::Parquet => {
+                let builder = ArrowBuilder::new(
+                    &connection_details.desc,
+                    DEFAULT_ARRAY_BUILDER_ITEM_CAPACITY,
+                    DEFAULT_ARRAY_BUILDER_DATA_CAPACITY,
+                )?;
+                Ok(ParquetUploader {
+                    desc: connection_details.desc,
+                    sdk_config: Arc::new(sdk_config),
+                    key_manager: S3KeyManager::new(sink_id, &connection_details.uri),
+                    batch,
+                    max_file_size: connection_details.max_file_size,
+                    next_file_index: 0,
+                    builder,
+                    row_group_size_bytes,
+                    arrow_builder_buffer_bytes,
+                    active_file: None,
+                    params,
+                })
+            }
+            _ => anyhow::bail!("Expected Parquet format"),
+        }
     }
-    async fn append_row(&mut self, _row: &Row) -> Result<(), anyhow::Error> {
-        anyhow::bail!("Parquet uploading is not yet implemented")
+
+    async fn append_row(&mut self, row: &Row) -> Result<(), anyhow::Error> {
+        self.builder.add_row(row)?;
+
+        if u64::cast_from(self.builder.row_size_bytes()) > self.arrow_builder_buffer_bytes {
+            self.flush_builder().await?;
+        }
+        Ok(())
     }
+
     async fn flush(&mut self) -> Result<(), anyhow::Error> {
-        anyhow::bail!("Parquet uploading is not yet implemented")
+        self.flush_builder().await?;
+        if let Some(active) = self.active_file.take() {
+            active.finish().await?;
+        }
```

Comment:
is this not handled in `flush_builder()`?

Reply:
No it isn't -- this `flush()` method is called when we want to complete the upload, so we need to `finish()` the active file if there is one. The `flush_builder()` method is just for flushing the `ArrowBuilder` to a file.

I think I will rename this method `finish()` as well, since we're essentially 'finishing' the upload completely

Reply:
actually this made me notice that we weren't running this `finish` call in a tokio task, will fix that too!